File: erlang/apauley/prices.erl
-module(prices).

-export([start/0,test/0]).

-record(product, {product_type,
                  product_code,
                  supplier_id,
                  description,
                  delivery_date,
                  cost_price,
                  unit_count,
                  sell_price,
                  sell_by_date}).

-define(PREMIUM_SUPPLIER_IDS, [204,219]).
-define(UNFRESH_SUPPLIER_IDS, [32,101]).

start() ->
  {ok, InputFile}  = file:open("../../produce.csv", [read]),
  {ok, OutputFile} = file:open("pricefile.txt", [write]),
  {ok, _Header} = file:read_line(InputFile),
  ok = process_file(InputFile, OutputFile),
  ok.

process_file(InputFile, OutputFile) ->
  case file:read_line(InputFile) of
    {ok, Line} ->
      process_file(InputFile, OutputFile, Line);
    eof ->
      ok
  end.

process_file(InputFile, OutputFile, Line) ->
  to_price_file(string:strip(Line, right, $\n), OutputFile),
  process_file(InputFile, OutputFile).

to_price_file(Line, OutputFile) ->
  Fields = csv:parse_line(Line),
  [SupplierID, ProductCode, Description, DeliveryDate, CostPrice, UnitCount] = Fields,
  Product = calculate_product(list_to_integer(SupplierID),list_to_integer(ProductCode),
                              Description, date_utils:parse_date(DeliveryDate),
                              list_to_integer(CostPrice), list_to_integer(UnitCount)),
  ok = write_pricefile(Product, OutputFile),
  ok.

calculate_product(SupplierID, ProductCode, Description, DeliveryDate, CostPrice, UnitCount) ->
  Product    = simple_product(SupplierID, ProductCode, Description, DeliveryDate, CostPrice, UnitCount),
  SellPrice  = sell_price(Product#product.product_type, SupplierID, CostPrice),
  SellByDate = sell_by_date(Product#product.product_type, SupplierID, DeliveryDate),
  Product#product{sell_price = SellPrice, sell_by_date=SellByDate}.

sell_price(ProductType, SupplierID, CostPrice) ->
  Cents = round(CostPrice + markup_cents(ProductType, SupplierID, CostPrice)),
  price_with_modifications(Cents, SupplierID).

price_with_modifications(BasePriceCents, SupplierID) ->
  MarkupCents1 = case lists:member(SupplierID, ?PREMIUM_SUPPLIER_IDS) of
                   true  -> cents_rounded_up_to_nearest_rand(BasePriceCents);
                   false -> BasePriceCents
                 end,
  MarkupCents2 = case lists:member(SupplierID, ?UNFRESH_SUPPLIER_IDS) of
                   true  -> MarkupCents1 - 200; % R2 discount
                   false -> MarkupCents1
                 end,
  max(0, MarkupCents2).

cents_rounded_up_to_nearest_rand(Cents) when is_integer(Cents) ->
  RandsPart = trunc(Cents/100),
  CentsPart = Cents rem 100,
  cents_rounded_up_to_nearest_rand(RandsPart, CentsPart).

cents_rounded_up_to_nearest_rand(RandsPart, _CentsPart=0) ->
  RandsPart * 100;
cents_rounded_up_to_nearest_rand(RandsPart, _CentsPart) ->
  (RandsPart + 1) * 100.

sell_by_date(ProductType, SupplierID, DeliveryDate) ->
  date_utils:date_add_days(DeliveryDate, shelf_days(SupplierID, ProductType)).

simple_product(SupplierID, ProductCode, Description, DeliveryDate, CostPrice, UnitCount) ->
  #product{product_type  = product_type(ProductCode),
           product_code  = ProductCode,
           supplier_id   = SupplierID,
           description   = Description,
           delivery_date = DeliveryDate,
           cost_price    = CostPrice,
           unit_count    = UnitCount}.

markup_cents(ProductType, SupplierID, CostPrice) ->
  Percentage = markup_percentage(ProductType, SupplierID),
  CostPrice * (Percentage/100.0).

markup_percentage(ProductType, SupplierID) ->
  markup_percentage(ProductType) + supplier_markup_percentage_modification(SupplierID).

markup_percentage(apple) ->
  40;
markup_percentage(banana) ->
  35;
markup_percentage(berry) ->
  55;
markup_percentage(_ProductType) ->
  50.

supplier_markup_percentage_modification(SupplierID) ->
  case lists:member(SupplierID, ?PREMIUM_SUPPLIER_IDS) of
    true  -> 10;
    false -> 0
  end.

shelf_days(_SupplierID=32, ProductType) ->
  shelf_days(ProductType) - 3;
shelf_days(_SupplierID, ProductType) ->
  shelf_days(ProductType).

shelf_days(apple)        -> 14;
shelf_days(banana)       -> 5;
shelf_days(_ProductType) -> 7.

product_type(ProductCode) when ((ProductCode >= 1100) and (ProductCode =< 1199)) ->
  apple;
product_type(ProductCode) when ((ProductCode >= 1200) and (ProductCode =< 1299)) ->
  banana;
product_type(ProductCode) when ((ProductCode >= 1300) and (ProductCode =< 1399)) ->
  berry.

label_sell_price(SellPriceCents) when is_integer(SellPriceCents) ->
  RandsPart = io_lib:format("~5.. B", [trunc(SellPriceCents/100)]),
  CentsPart = io_lib:format("~2..0B", [SellPriceCents rem 100]),
  "R"++ lists:flatten(RandsPart ++ "." ++ CentsPart).

write_pricefile(Product, OutputFile) ->
  LabelSellPrice = label_sell_price(Product#product.sell_price),
  LabelSellByDate  = date_utils:date_to_string(Product#product.sell_by_date),
  LabelDescription = string:substr(Product#product.description, 1, 31),
  Line = LabelSellPrice ++ LabelSellByDate ++ LabelDescription ++ "\n",
  Lines = [Line || _Count <- lists:seq(1,Product#product.unit_count)],
  ok = file:write(OutputFile, Lines).

test() ->
  %% http://armstrongonsoftware.blogspot.com/2009/01/micro-lightweight-unit-testing.html
  test_sell_price(),
  test_sell_price_premium_supplier(),
  test_sell_price_unfresh_supplier(),
  test_label_sell_price(),
  ok.

test_sell_price() ->
  1400 = sell_price(_ProductType=apple, SupplierID=0, CostPrice=1000),
  1350 = sell_price(banana, SupplierID, CostPrice),
  1550 = sell_price(berry,  SupplierID, CostPrice),
  1500 = sell_price(other,  SupplierID, CostPrice),
  0    = sell_price(other,  SupplierID, 0),
  ok.

test_sell_price_premium_supplier() ->
  [SupplierID|_] = ?PREMIUM_SUPPLIER_IDS,
  1600 = sell_price(_ProductType=apple, SupplierID, _CostPrice=1010),
  1500 = sell_price(_ProductType=apple, SupplierID, 1000),
  ok.

test_sell_price_unfresh_supplier() ->
  [SupplierID|_] = ?UNFRESH_SUPPLIER_IDS,
  1200 = sell_price(_ProductType=apple, SupplierID, _CostPrice=1000),
  70   = sell_price(banana, SupplierID, 200),
  0    = sell_price(berry,  SupplierID, 0),
  ok.

test_label_sell_price() ->
  "R   17.23" = label_sell_price(_Cents=1723),
  "R  123.04" = label_sell_price(12304),
  "R99999.99" = label_sell_price(9999999),
  "R    1.23" = label_sell_price(123),
  "R    1.20" = label_sell_price(120),
  "R    0.23" = label_sell_price(23),
  "R    0.01" = label_sell_price(1),
  "R    0.00" = label_sell_price(0),
  ok.

File: ruby/apauley/prices.rb
#!/usr/bin/env ruby

require 'csv'
require 'date'

PremiumSupplierIDs = [204,219]
UnfreshSupplierIDs = [32,101]

class Product
  def self.create(supplier, product_code, description, delivery_date, cost_price, unit_count)
    if product_code >= 1000 and product_code <= 1999
      Fruit.create(supplier, product_code, description, delivery_date, cost_price, unit_count)
    else
      raise "Bad product code: #{product_code} #{description}"
    end
  end
  
  def initialize(supplier, product_code, description, delivery_date, cost_price, unit_count)
    @supplier      = supplier
    @product_code  = product_code
    @description   = description
    @delivery_date = delivery_date
    @cost_price    = cost_price
    @unit_count    = unit_count
  end

  def shelf_days
    7
  end

  def sell_by_date
    @delivery_date + shelf_days + @supplier.shelf_days_modifier
  end

  def product_markup_percentage
    50
  end

  def markup_percentage
    product_markup_percentage + @supplier.markup_percentage_modifier
  end

  def markup
    @cost_price*(markup_percentage/100.0)
  end

  def sell_price
    base_sell_price = @cost_price + markup
    @supplier.modify_sell_price(base_sell_price)
  end

  def sell_price_in_rands
    sell_price/100.0
  end

  def label_sell_price
    'R' + "% 8.2f" % sell_price_in_rands
  end

  def label_sell_by_date
    sell_by_date.strftime('%Y/%m/%d')
  end

  def label_description
    @description[0, 31]
  end
  
  def write_pricefile(file)
    @unit_count.times {file.puts label_sell_price + label_sell_by_date + label_description}
  end
end

class Fruit < Product
  def self.create(supplier, product_code, description, delivery_date, cost_price, unit_count)
    if (1100..1199).include?(product_code)
      Apple.new(supplier, product_code, description, delivery_date, cost_price, unit_count)
    elsif (1200..1299).include?(product_code)
      Banana.new(supplier, product_code, description, delivery_date, cost_price, unit_count)
    elsif (1300..1399).include?(product_code)
      Berry.new(supplier, product_code, description, delivery_date, cost_price, unit_count)
    else
      raise "Bad Fruit product code: #{product_code} #{description}" 
    end
  end
end

class Apple < Fruit
  def product_markup_percentage
    40
  end
  def shelf_days
    14
  end
end

class Banana < Fruit
  def product_markup_percentage
    35
  end
  def shelf_days
    5
  end
end

class Berry < Fruit
  def product_markup_percentage
    55
  end
end

class Supplier
  def self.create(supplier_id)
    if PremiumSupplierIDs.include?(supplier_id)
      PremiumSupplier.new(supplier_id)
    elsif UnfreshSupplierIDs.include?(supplier_id)
      UnfreshSupplier.new(supplier_id)
    else
      self.new(supplier_id)
    end
  end
  
  def initialize(supplier_id)
    @supplier_id = supplier_id
  end

  def id
    @supplier_id
  end

  def shelf_days_modifier
    0
  end

  def modify_sell_price(sell_price)
    sell_price
  end

  def markup_percentage_modifier
    0
  end
end

class PremiumSupplier < Supplier
  def markup_percentage_modifier
    10
  end

  def modify_sell_price(sell_price)
    (sell_price/100.0).ceil * 100
  end
end

class UnfreshSupplier < Supplier
  def shelf_days_modifier
    -3
  end

  def modify_sell_price(sell_price)
    [0, sell_price - 200].max
  end
end

def to_price_file(line, file)
  CSV.parse(line) do |row|
    supplier_id   = Integer(row[0])
    product_code  = Integer(row[1])
    description   = row[2]
    delivery_date = Date.parse(row[3])
    cost_price    = Integer(row[4])
    unit_count    = Integer(row[5])
    
    supplier = Supplier.create(supplier_id)
    product  = Product.create(supplier, product_code, description, delivery_date, cost_price, unit_count)
    product.write_pricefile(file)
  end
end

def main()
  inputfile  = File.new('../../produce.csv', 'r')
  outputfile = File.new('pricefile.txt', 'w')

  inputfile.each_line("\n") do |line|
    if inputfile.lineno > 1
      to_price_file(line, outputfile)
    end
  end
end

main()
